ruma-common/src/push/condition.rs

```diff
@@ -42,13 +52,273 @@ pub enum PushCondition {
+    ///
+    /// * `event` - The flattened JSON representation of a room message event.
+    /// * `context` - The context of the room at the time of the event.
+    pub fn applies(&self, event: &FlattenedJson, context: &PushConditionRoomCtx) -> bool {
+        match self {
+            Self::EventMatch { key, pattern } => {
+                let value = match key.as_str() {
+                    "room_id" => &context.room_id,
+                    _ => match event.get(key) {
+                        Some(v) => v,
+                        None => return false,
+                    },
+                };
+
+                value.matches_pattern(pattern, key == "content.body")
+            }
+            Self::ContainsDisplayName => {
+                let value = match event.get("content.body") {
+                    Some(v) => v,
+                    None => return false,
+                };
+
+                value.matches_pattern(&context.user_display_name, true)
+            }
+            Self::RoomMemberCount { is } => is.contains(&context.member_count),
+            Self::SenderNotificationPermission { key } => {
+                let sender_id = match event.get("sender") {
+                    Some(v) => match UserId::try_from(v.as_str()) {
+                        Ok(u) => u,
+                        Err(_) => return false,
+                    },
+                    None => return false,
+                };
+
+                let sender_level = context
+                    .users_power_levels
+                    .get(&sender_id)
+                    .unwrap_or(&context.default_power_level);
+
+                match context.notification_power_levels.get(key) {
+                    Some(l) => sender_level >= l,
+                    None => false,
+                }
+            }
+        }
+    }
+}
+
+/// The context of the room associated to an event to be able to test all push conditions.
+#[derive(Clone, Debug)]
+pub struct PushConditionRoomCtx {
+    /// The roomId of the room.
+    pub room_id: String,
+
+    /// The number of members in the room.
+    pub member_count: UInt,
+
+    /// The display name of the current user in the room.
+    pub user_display_name: String,
+
+    /// The power levels of the users of the room.
+    pub users_power_levels: BTreeMap<UserId, Int>,
+
+    /// The default power level of the users of the room.
+    pub default_power_level: Int,
+
+    /// The notification power levels of the room.
+    pub notification_power_levels: NotificationPowerLevels,
+}
+
+/// Additional functions for character matching.
+pub trait CharExt {
+    /// Whether or not this char can be part of a word.
+    fn is_word_char(&self) -> bool;
+}
+
+impl CharExt for char {
+    fn is_word_char(&self) -> bool {
+        self.is_alphanumeric() || self == &'_'
+    }
+}
+
+/// Additional functions for string matching.
+pub trait StrExt {
+    /// Get the length of the char at `index` in `string`. The byte index must correspond to
```

Comment:
```suggestion
    /// Get the length of the char at `index` in `self`. The byte index must correspond to
```

Same below.

Reply:
Sorry I forgot to update some docs. Is it necessary to mention `self`? I'd just remove it:
```suggestion
    /// Get the length of the char at `index`. The byte index must correspond to
```

Reply:
Sure, that works too.